File: Ruby/simpsons_rule.rb
# simpsons_rule.rb
# Module for Simpson's Rule Integration in Ruby

module SimpsonsRule
  # Method to calculate the integral using Simpson's Rule
  # @param f [Proc] The function to integrate
  # @param a [Float] The lower limit of integration
  # @param b [Float] The upper limit of integration
  # @param n [Integer] The number of sub-intervals (must be even)
  # @return [Float] The approximate value of the integral

  def self.calculate(f, a, b, n)
    # Ensure `n` is even
    n += 1 if n.odd?

    h = (b - a) / n.to_f
    result = f.call(a) + f.call(b)

    (1...n).each do |i|
      x = a + i * h
      result += (i.even? ? 2 : 4) * f.call(x)
    end

    result * h / 3
  end
end

File: R/simpsons_rule.R
# simpsons_rule.R
# Simpson's Rule Integration Function in R

# Function to calculate the integral using Simpson's Rule
# Args:
#   f: Function to integrate
#   a: Lower limit of integration
#   b: Upper limit of integration
#   n: Number of intervals (must be even)
# Returns:
#   Approximate integral value

simpsons_rule <- function(f, a, b, n) {
  # Ensure n is even
  if (n %% 2 != 0) {
    n <- n + 1
  }
  
  h <- (b - a) / n
  result <- f(a) + f(b)
  
  for (i in 1:(n - 1)) {
    x <- a + i * h
    if (i %% 2 == 0) {
      result <- result + 2 * f(x)
    } else {
      result <- result + 4 * f(x)
    }
  }
  
  result * h / 3
}
